internal/backend/remote-state/inmem/backend.go

```diff
@@ -48,19 +50,57 @@ func Reset() {
 
 // New creates a new backend for Inmem remote state.
 func New() backend.Backend {
+	if os.Getenv("TF_INMEM_TEST") != "" {
```

Comment:
A choice that I needed to consider here was either to:
* Make the `inmem` backend have test-specific schemas.
* Or, add a new in-memory backend that was only available for use in the context of testing.

Currently the `inmem` backend is available for end users to use if they know it exists (it is not in any docs), but it was intended as a test resource. Making the `inmem` backend unavailable to end users and only available during tests is a breaking change. We could do that in future, but for now I think making parts of the inmem backend test-only makes sense, instead of creating a new backend that uses 99% of the same code.

Reply:
I think it'd be great to have this comment somewhere in this code for our future selves.

Reply:
Good point - I've included a version of my comment in [7a93fdc](https://github.com/hashicorp/terraform/pull/36919/commits/7a93fdc0276ba9ac5f7169afb6ad91bec03765a4)